docs/jax-101/06-parallelism.md

```diff
@@ -0,0 +1,423 @@
+
+```{code-cell} ipython3
+:id: YSCgHguTSdGW
+:outputId: a8bf28df-3747-4d49-e340-b7696cf0c27d
+
+type(replicated_params.weight)
+```
+
++++ {"id": "90VtjPbeY-hD"}
+
+The params will become a ShardedDeviceArray when they are returned by our pmapped `update()` (see further down).
+
++++ {"id": "eGVKxk1CV-m1"}
+
+We do the same to the data:
+
+```{code-cell} ipython3
+:id: vY61QJoFWCII
+:outputId: f436a15f-db97-44cc-df33-bbb4ff222987
+
+def split(arr):
+  """Splits the first axis of `arr` evenly across the number of devices."""
+  return arr.reshape(n_devices, arr.shape[0] // n_devices, *arr.shape[1:])
+
+# Reshape xs and ys for the pmapped `update()`.
+x_split = split(xs)
+y_split = split(ys)
+
+type(x_split)
+```
+
++++ {"id": "RzfJ-oK5WERq"}
+
+The data is just a reshaped vanilla NumPy array. Hence, it cannot be anywhere but on the host, as NumPy runs on CPU only. Since we never modify it, it will get sent to the device at each `update` call, like in a real pipeline where data is typically streamed from CPU to the device at each step.
+
+```{code-cell} ipython3
+:id: atOTi7EeSQw-
+:outputId: c8daf141-63c4-481f-afa5-684c5f7b698d
+
+def type_after_update(name, obj):
+  print(f"after first `update()`, `{name}` is a", type(obj))
+
+# Actual training loop.
+for i in range(1000):
+
+  # This is where the params and data gets communicated to devices:
+  replicated_params, loss = update(replicated_params, x_split, y_split)
+
+  # The returned `replicated_params` and `loss` are now both ShardedDeviceArrays,
+  # indicating that they're on the devices.
+  # `x_split`, of course, remains a NumPy array on the host.
+  if i == 0:
+    type_after_update('replicated_params.weight', replicated_params.weight)
+    type_after_update('loss', loss)
+    type_after_update('x_split', x_split)
+
+  if i % 100 == 0:
+    # Note that loss is actually an array of shape [num_devices], with identical
+    # entries, because each device returns its copy of the loss.
+    # So, we take the first element to print it.
+    print(f"Step {i:3d}, loss: {loss[0]:.3f}")
+
+
+# Plot results.
+
+# Like the loss, the leaves of params have an extra leading dimension,
+# so we take the params from the first device.
+params = jax.device_get(jax.tree_map(lambda x: x[0], replicated_params))
+```
+
+```{code-cell} ipython3
+:id: rvVCACv9UZcF
+:outputId: 5c472d0f-1236-401b-be55-86e3dc43875d
+
+import matplotlib.pyplot as plt
+plt.scatter(xs, ys)
+plt.plot(xs, params.weight * xs + params.bias, c='red', label='Model Prediction')
+plt.legend()
+plt.show()
+```
+
++++ {"id": "4wFJcqbhbn81"}
+
+## Aside: hosts and devices in JAX
+
+When running on TPU, the idea of a 'host' becomes important. A host is the CPU that manages several devices. A single host can only manage so many devices (usually 8, but with some tricks can be extended to 16), so when running very large parallel programs, multiple hosts are needed, and some finesse is required to manage them.
```

Comment:
Let's not mention the part about extending to 16 devices, you can only do that internally. I would just say "(usually 8)".